ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,313 @@
+            return rootHandle;
+        }
+    }
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    UpdateCacheForVnode(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHash,
+        vnodeVid,
+        invalidateEntry,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (nullptr != perfTracer)
+    {
+        perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+    }
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode)
+{
+    uintptr_t vnodeAddress = reinterpret_cast<uintptr_t>(vnode);
+    return (vnodeAddress >> 3) % s_entriesCapacity;
+}
+
+KEXT_STATIC bool TryGetVnodeRootFromCache(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    bool rootFound = false;
+    rootHandle = RootHandle_None;
+
+    RWLock_AcquireShared(s_entriesLock);
+    {
+        uintptr_t vnodeIndex;
+        if (TryFindVnodeIndex_Locked(vnode, vnodeHash, /*out*/ vnodeIndex))
+        {
+            if (vnode == s_entries[vnodeIndex].vnode && vnodeVid == s_entries[vnodeIndex].vid)
+            {
+                rootFound = true;
+                rootHandle = s_entries[vnodeIndex].virtualizationRoot;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_entriesLock);
+    
+    return rootFound;
+}
+
+KEXT_STATIC void UpdateCacheForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        if (!TryUpdateEntryToLatest_ExclusiveLocked(
+                perfTracer,
+                cacheMissFallbackFunctionCounter,
+                cacheMissFallbackFunctionInnerLoopCounter,
+                vnode,
+                context,
+                vnodeHash,
+                vnodeVid,
+                invalidateEntry,
+                /* out */ rootHandle))
+        {
+            // TryUpdateEntryToLatest_ExclusiveLocked can only fail if the cache is full
+            
+            perfTracer->IncrementCount(PrjFSPerfCounter_CacheFullCount, true /*ignoreSampling*/);
+        
+            InvalidateCache_ExclusiveLocked();
```

Comment:
To refresh my memory if the cache is full, we just reset it correct?  I.E. no growing to worry about yet.  That may come in the future.

Reply:
That's the current design/implementation.  

With the simplification to the locking scheme it might not be too bad to support "reset + grow" rather than a simple reset (i.e. allocate more space but don't migrate existing entries).  

I've added that investigation to my TODO list.

(I'm not sure if we'd still want to enforce some sort of upper-bound on the size of the cache though, @pmj what are your thoughts?)

Reply:
Filed #924 for this.